src/main/java/com/mongodb/spark/sql/connector/read/partitioner/AutoBucketPartitioner.java

```diff
@@ -0,0 +1,337 @@
+      LOGGER.warn("Unable to get collection stats (collstats) returning a single partition.");
+      return SINGLE_PARTITIONER.generatePartitions(readConfig);
+    }
+
+    double avgObjSizeInBytes =
+        storageStats.get("avgObjSize", new BsonInt32(0)).asNumber().doubleValue();
+    double numDocumentsPerPartition = Math.floor(partitionSizeInBytes / avgObjSizeInBytes);
+
+    BsonDocument usersMatchQuery =
+        PartitionerHelper.matchQuery(readConfig.getAggregationPipeline());
+    long count;
+    if (usersMatchQuery.isEmpty() && storageStats.containsKey("count")) {
+      count = storageStats.getNumber("count").longValue();
+    } else {
+      count = readConfig.withCollection(coll -> coll.countDocuments(
+          usersMatchQuery, new CountOptions().comment(readConfig.getComment())));
+    }
+
+    if (numDocumentsPerPartition >= count) {
+      LOGGER.info(
+          "Fewer documents ({}) than the calculated number of documents per partition ({}). Returning a single partition",
+          count,
+          numDocumentsPerPartition);
+      return SINGLE_PARTITIONER.generatePartitions(readConfig);
+    }
+
+    int numberOfBuckets = (int) Math.ceil(count / numDocumentsPerPartition);
+    int numberOfSamples = (int) Math.ceil(samplesPerPartition * numDocumentsPerPartition);
+
+    List<BsonDocument> buckets =
+        readConfig.withCollection(coll -> coll.aggregate(createBucketAutoPipeline(
+                usersMatchQuery,
+                partitionFieldList,
+                partitionProjectionKey,
+                numberOfSamples,
+                numberOfBuckets))
+            .allowDiskUse(readConfig.getAggregationAllowDiskUse())
+            .comment(readConfig.getComment())
+            .into(new ArrayList<>()));
+
+    return createMongoInputPartitions(
+        buckets,
+        readConfig.getAggregationPipeline(),
+        partitionFieldList,
+        partitionProjectionKey,
+        getPreferredLocations(readConfig));
+  }
+
+  /**
+   * Creates the $sample and $bucketAuto aggregation pipeline used for determining the partition bounds.
+   *
+   * @param usersMatchQuery the match part of a user supplied aggregation pipeline or an empty document
+   * @param partitionFieldList the fields to partition the collection by
+   * @param partitionProjectionKey the partition projection key only used if there are multiple partition fields
+   * @param numberOfSamples the number of samples
+   * @param numberOfBuckets the number of buckets
+   * @return the pipeline
+   */
+  @VisibleForTesting
+  static List<BsonDocument> createBucketAutoPipeline(
+      final BsonDocument usersMatchQuery,
+      final List<String> partitionFieldList,
+      final String partitionProjectionKey,
+      final int numberOfSamples,
+      final int numberOfBuckets) {
+
+    List<BsonDocument> pipeline = new ArrayList<>();
+    if (!usersMatchQuery.isEmpty()) {
+      pipeline.add(new BsonDocument("$match", usersMatchQuery));
```

Comment:
This line rises two questions:

1. Why do we need to "wrap" `usersMatchQuery` in `$match`, given that it's already a `$match` itself?
2. Why can't we just append the user pipeline here to `pipeline` (as we do in `createPartitionPipeline`), and instead have to have the user pipeline wrapped in `$match`?

Reply:
This is because it was unwrapped originally, so it could be used in the `collection.countDocuments` - will rename to clarify.